### PR TITLE
Fork Function

### PR DESCRIPTION
**Pull request by farizrahman4u**
Splits the data flow to two or more streams in Sequential model.

### CODE REVIEW

**Comment by farizrahman4u** [comment]
API usage:

```
root = Sequential()
root.add(Dense(input_dim=10, output_dim=10))

forks=root.fork(n=3) #Split to 3 forks, default value is 2

forks[0].add(Dense(5))

forks[1].add(Dense(10))
forks[1].add(Dense(5))

forks[2].add(LSTM(5))

model = Sequential()
model.add(Merge(forks,mode='sum')) # Merge the forks
model.compile(omptimizer='sgd', loss='mse')

```

**Comment by farizrahman4u** [comment]
**API usage**:

```
root = Sequential()
root.add(Dense(input_dim=10, output_dim=10))

forks = root.fork(n=3) #Split to 3 forks, default value is 2
left, middle, right = forks

left.add(Dense(5))

middle.add(Dense(10))
middle.add(Dense(5))

right.add(LSTM(5))

model = Sequential()
model.add(Merge(forks,mode='sum')) # Merge the forks
model.compile(omptimizer='sgd', loss='mse')
```

**Comment by sergeyf** [comment]
This has the same kind of functionality as is native in the Graph model?

**Comment by farizrahman4u** [comment]
@sergeyf  I have updated my first comment.

**Comment by sergeyf** [comment]
Great, thank you!

**Comment by farizrahman4u** [comment]
@sergeyf  What do you think about it ? From a UX perspective?

**Comment by sergeyf** [comment]
It looks pretty intuitive. 

One thing I am regularly want to do is a fork but keeping the layers linked in all branches. So, for something like a siamese network, have two input vectors go down separate branches, but have those branches be identical throughout the training in a native way. It would be cool if this was possible with a `fork` but that's probably way more work =)

**Comment by farizrahman4u** [comment]
@sergeyf  Could you please elaborate? Maybe some pseudocode?

**Comment by sergeyf** [comment]
Sure. Let's say I have pairs of sentences represented as bags-of-words, and I have some measure of similarity between them. I'd like to train a siamese nn that takes pairs and learns whether they are similar or not (binary classification).

So, how to do this? I have to use `Graph`

```
model = Graph()

# inputs
model.add_input(name='first_sentence', ndim=2)
model.add_input(name='second_sentence', ndim=2)

model.add_node(Dense(input_dim=d, output_dim=100, activation='sigmoid'),name='top_branch', input='first_sentence')

model.add_node(Dense(input_dim=d, output_dim=100, activation='sigmoid'),name='bottom_branch', input='second_sentence')

model.add_node(Dense(output_dim=1,activation='softmax'),name='output_dense',inputs=['top_branch','bottom_branch'])
```

Except I'd like `top_branch` to always have the same parameters as `bottom_branch` -> I don't want to learn two encodings. 

Now that I think about it, this is different from `fork`. With `fork`, the goal is to do different operations simultaneously to the same data, but with what I want to do, the goal is to do the same operation to different data simultaneously. 

`fork` does this:
```
      /---------
-----|
      \---------
```

But I need this kind of branching:
```
----------\
           |---------
----------/
```

Where the top and bottom branch are shared/identical.

**Comment by farizrahman4u** [comment]
@sergeyf  I am on it!

**Comment by sergeyf** [comment]
Ah, so concatenate along the first axis and then do a reshape after the Dense?

**Comment by farizrahman4u** [comment]
@sergeyf  Yeah.. but i still think there is a need for a "siamese" layer.

**Comment by farizrahman4u** [comment]
@sergeyf  No need.. I just did a google search for siamese networks and there are many of them. It is also there in [Caffe](http://caffe.berkeleyvision.org/gathered/examples/siamese.html). So as I said, I will   be back in a couple of hours.

**Comment by farizrahman4u** [comment]
OK, that took less than hour! Checkout #928 .

**Comment by sergeyf** [comment]
Thanks! I am not sure - I'll have to give to try to actually use it and then I will get back to you. I appreciate the time you've given this.

**Comment by farizrahman4u** [comment]
@sergeyf  I have written a fully sequential implementation of Siamese network here: #928. I would love to hear your feedback about this.

**Comment by sergeyf** [comment]
Very cool! And very fast =)

It looks very flexible and useful. The `get_siamese_heads` function feels a bit tacked on, but I am not sure how else to get it to do what it's doing without some changes to the base `model` class. I assume you are trying to avoid that? 

Also, how does this fit in with the `Graph` api? Probably some of the more regular contributors to Keras would have a lot more to say, but as it is I would find immediate use for it, and so would many others I bet.

**Comment by farizrahman4u** [comment]
@sergeyf  As we are talking about `Siamese` now, we could move this conversation there.

**Comment by sergeyf** [comment]
Sure, let's do that. I'll paste my last comment.

**Comment by farizrahman4u** [comment]
@sergeyf  Yes, please..

**Comment by fchollet** [comment]
> What do you think about it ? From a UX perspective?

Before implementing `Graph`, we investigated implementing `Fork` (with a somewhat different API, mirroring `Merge`), something like:

```python
head1, head2 = Fork(sequential_model, n=2)
```

The reason why I am pessimistic about it is that it does not offer true branching support: you have to merge back your heads back into a single output for compilation. This is rarely what you want when you need to fork a data stream. 

Meanwhile, adding support for true branching support to `Sequential` would be very heavy from a UX perspective and would largely defeat the point of having separate `Sequential` and `Graph` models. `Graph` was introduced as a simpler, more elegant and more general solution to these problems, while `Merge` was maintained anyway because it still made certain types of models a bit simpler.

**Comment by dbonadiman** [comment]
I agree with @fchollet all this changes meant  to implement simple graph based models in Sequential model are not worth.
We should instead focus our effort into build example of the graph model usage that is much more powerful in that kinds of architecture. All the shared layer stuff is straight forward in a Graph architecture and it can be implemented with simple knowledge of the library. Conversely it needs much more effort in Sequential layers. 

What i'm saying is that adding these mode would increase the sparsity of the library or even break existing code that worked in older version.

**Comment by farizrahman4u** [comment]
@dbonadiman 
> What i'm saying is that adding these mode would increase the sparsity of the library or even break existing code that worked in older version.

Wrong. It wont break any existing code.

**Comment by farizrahman4u** [comment]
@fchollet 
>  it does not offer true branching support: you have to merge back your heads back into a single output for compilation. This is rarely what you want when you need to fork a data stream.

Multiple outputs are not the only intentions of forking. In many models, which look almost sequential, i have seen minor forks which are merged immediately.

In [this paper](http://research.microsoft.com/pubs/246720/rnn_em.pdf) for instance, `k(t)` and `beta(t)` are learned from `h(t)`, but they are immediately merged in Eqn12.

**Comment by fchollet** [comment]
One more thing: as per e.g. [this comment](https://groups.google.com/forum/#!topic/keras-users/SzQeBgXA70k), and following contributor guidelines, these changes should be documented.

**Comment by fchollet** [comment]
We won't merge this. I think we should focus our efforts on improving the UX of the `Graph` API.
